testpartitioning/testFiltering.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package testpartitioning
+
+import (
+	"os"
+	"strconv"
+	"testing"
+)
+
+// PartitionTest checks if the current partition should run this test, and skips it if not.
+func PartitionTest(t *testing.T) {
+	pt, found := os.LookupEnv("PARTITION_TOTAL")
+	if !found {
+		return
+	}
+	partitions, err := strconv.Atoi(pt)
+	if err != nil {
+		return
+	}
+	pid := os.Getenv("PARTITION_ID")
+	partitionID, err := strconv.Atoi(pid)
+	if err != nil {
+		return
+	}
+	name := t.Name()
+	nameNumber := stringToUint64(name)
+	if nameNumber%uint64(partitions) != uint64(partitionID) {
+		t.Skip()
+	}
+}
+
+func stringToUint64(str string) uint64 {
+	sum := uint64(0)
+	for _, x := range str {
+		sum += uint64(x)
+	}
+	return sum
```

Comment:
Sorry for this comment, this obviously isn't critical.

Technically the test name isn't going to give you a uniformly distributed partition key, so you would need to hash this:
```suggestion
	h := fnv.New64a()
	h.Write([]byte(s))
	return h.Sum64()
```

This isn't perfect either, since there are tests with the same name in different packages:
```
~$ rg "func Test"| cut -f 2 -d ' '|sort|wc -l
1360

~$ rg "func Test"| cut -f 2 -d ' '|sort|uniq|wc -l
1344
```

Reply:
oh interesting, for a more even distribution you could get the filename & line number of the caller too by `runtime.Caller(1)`. Also, passing the `pc` value it returns to `runtime.FuncForPC` could get you a version of the calling function name and package name, here's an example: https://play.golang.org/p/tEW91-qEdqR

Reply:
@cce I was thinking about something similar, but wouldn't that give the line/file of the helper function instead of the test function? I guess hashing a random value would work too (unless the partitions should be deterministic).

Reply:
`Caller(N)` will go N levels up the stack, so it will get the test function instead of the helper function — Caller() and FuncForPC() are sometimes used in logging/metrics/tracing helper functions to grab names like this for capturing monitoring data.

Deterministic seems good IMO, especially for re-running tests

Reply:
I gave a try at implementing @winder's idea along with the `runtime.Caller(1)` in https://github.com/cce/go-algorand/commit/0cf3b2d1a97fba19fb7f225c17487a3b6758f209 CC @algonautshant